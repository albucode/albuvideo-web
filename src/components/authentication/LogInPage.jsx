import React from "react";
import styled from '@emotion/styled'
import theme  from "../../theme/theme"
import { Box, Text } from "@chakra-ui/react";


export const LogInPage = () => {
  return (
    <PageBackground >
      <Image>

      </Image>
        <FormBackground>
          <FormContainer>
            <Header>Login To Your Account</Header>
          </FormContainer>

        </FormBackground>
    </PageBackground>
  );
}

const Image = styled(Box)`
width: 70%;
`

const PageBackground = styled(Box)`
background-color: ${theme.colors.lightBlue};
height: 100vh;
display: flex;
`

const FormContainer = styled(Box)`
padding-top: 25%;
`

const FormBackground = styled(Box)`
background-color: white;
border-radius: 12px; 
width: 30%;
height: 90%;
margin-top: 32px;
margin-right: 56px;
`

const Header = styled(Text)`
font-weight: 700;
font-size: 24px;
text-align: center;
`
