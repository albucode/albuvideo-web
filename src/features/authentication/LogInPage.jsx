import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router-dom";
import styled from "@emotion/styled";
import theme from "../../theme/theme";
import {
  Box,
  Text,
  FormControl,
  FormLabel,
  Input,
  Button,
  Image,
  Center,
} from "@chakra-ui/react";
import { Session } from "../../api/requests";
import { loadUser } from "./userSlice";
import { displayErrorAlert, loadErrorMessage } from "../shared/errorAlertSlice";
import ErrorAlert from "../shared/ErrorAlert";

export const LogInPage = () => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  const { errorMessage, displayErrorMessage } = useSelector(
    (state) => state.errorAlert
  );

  const dispatch = useDispatch();
  const history = useHistory();

  const createSession = (event) => {
    event.preventDefault();

    const requestBody = {
      user: {
        email,
        password,
      },
    };

    Session.create(requestBody).then((response) => {
      if (!response.error) {
        dispatch(loadUser(response));
        history.push("/dashboard");
      } else {
        dispatch(loadErrorMessage(response.error));
        dispatch(displayErrorAlert(true));
      }
    });
  };

  return (
    <PageBackground>
      <ImageContainer>
        <Center height="100vh">
          <Image src="computer-work.png" />
        </Center>
      </ImageContainer>
      <FormBackground>
        <FormContainer>
          <Header>Login To Your Account</Header>
          <form onSubmit={createSession}>
            <FormItem>
              <Label>Email address</Label>
              <Input
                type="email"
                onChange={(event) => setEmail(event.target.value)}
              />
            </FormItem>
            <FormItem>
              <Label>Password</Label>
              <Input
                type="password"
                onChange={(event) => setPassword(event.target.value)}
              />
            </FormItem>
            {displayErrorMessage && <ErrorAlert message={errorMessage} />}
            <LogInButton type="submit">Log In</LogInButton>
          </form>
        </FormContainer>
      </FormBackground>
    </PageBackground>
  );
};

const PageBackground = styled(Box)`
  background-color: ${theme.colors.lightBlue};
  height: 100vh;
  display: flex;
`;

const FormBackground = styled(Box)`
  background-color: white;
  border-radius: 12px;
  width: 30%;
  height: 90%;
  margin-top: 32px;
  margin-right: 56px;
`;

const FormContainer = styled(Box)`
  padding-top: 25%;
  padding-right: 20%;
  padding-left: 20%;
`;

const Header = styled(Text)`
  font-weight: 700;
  font-size: 24px;
  text-align: center;
  padding-bottom: 48px;
`;

const Label = styled(FormLabel)`
  color: ${theme.colors.black};
  font-size: 18px;
`;

const FormItem = styled(FormControl)`
  padding-bottom: 24px;
`;

const ImageContainer = styled(Box)`
  width: 70%;
`;

const LogInButton = styled(Button)`
  background-color: blue;
  color: white;
  width: 100%;
  margin-top: 24px;
`;
