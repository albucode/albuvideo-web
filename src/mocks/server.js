import { setupServer } from "msw/node";
import { rest } from "msw";

import { baseUrl } from "../config";

export const handlers = [
  rest.post(`${baseUrl}/users/sign_in`, (req, res, ctx) => {
    return res(
      ctx.json({ error: "You need to sign in or sign up before continuing." })
    );
  }),
];

// This configures a request mocking server with the given request handlers.
export const server = setupServer(...handlers);
