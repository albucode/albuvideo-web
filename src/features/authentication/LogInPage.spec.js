import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";

import { LogInPage } from "./LogInPage";
import userReducer from "../authentication/userSlice";
import errorAlertReducer from "../shared/errorAlertSlice";
import { server } from "../../mocks/server";

const fakeRenderComponent = () => {
  const store = configureStore({
    reducer: {
      user: userReducer,
      errorAlert: errorAlertReducer,
    },
    preloadedState: {
      user: {
        email: "",
        emailInitial: "",
        isLoggedIn: false,
      },
      errorAlert: {
        errorMessage: "",
        displayErrorMessage: false,
      },
    },
  });

  const { container } = render(
    <Provider store={store}>
      <LogInPage />
    </Provider>
  );
  return { container };
};

describe("Renders LogInPage components", () => {
  beforeEach(() => {
    fakeRenderComponent();
  });

  it("renders header", () => {
    expect(screen.getByTestId("header")).toHaveTextContent(
      "Login To Your Account"
    );
  });

  it("renders email field", () => {
    expect(screen.getByTestId("email-field")).toHaveTextContent(
      "Email address"
    );
  });

  it("renders password field", () => {
    expect(screen.getByTestId("password-field")).toHaveTextContent("Password");
  });

  it("renders a login button", () => {
    expect(screen.getByTestId("login-button")).toHaveTextContent("Log In");
  });

  it("matches snapshot", () => {
    const { container } = fakeRenderComponent();
    expect(container).toMatchSnapshot();
  });
});

describe("Form behaviour", () => {
  beforeAll(() => server.listen());

  afterEach(() => server.resetHandlers());

  afterAll(() => server.close());

  it("renders error message", async () => {
    // await act (async () => {
    //   fireEvent.change(screen.getByTestId("email-field"), {
    //     target: {email: ''},
    //   });
    //
    //   fireEvent.change(screen.getByTestId("password-field"), {
    //     target: {password: ''},
    //   });
    // });

    fakeRenderComponent();
    fireEvent.submit(screen.getByTestId("form"));
    expect(
      await screen.findByText(
        /You need to sign in or sign up before continuing/i
      )
    ).toBeInTheDocument();
  });

  // it('should submit when form inputs contain text', async () => {
  //   const { getByTestId, queryByText } = render(<Login/>)
  //
  //   await act(async () => {
  //     fireEvent.change(screen.getByLabelText(/username/i), {
  //       target: {value: 'shaquille'},
  //     });
  //
  //     fireEvent.change(screen.getByLabelText(/password/i), {
  //       target: {value: 'oatmeal'},
  //     })
  //   });
  //
  //   await act (async () => {
  //     fireEvent.submit(getByTestId('form'))
  //   });
  //
  //   expect(queryByText("User Name is required")).not.toBeInTheDocument();
  //   expect(queryByText("Password is required")).not.toBeInTheDocument();
  // });
});
