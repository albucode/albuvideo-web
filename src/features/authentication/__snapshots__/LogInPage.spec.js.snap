// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders LogInPage components matches snapshot 1`] = `
<div>
  <div
    class="css-1c4hr0u"
  >
    <div
      class="css-1mfqb2z"
    >
      <div
        class="css-h74a88"
      >
        <img
          class="chakra-image__placeholder css-0"
        />
      </div>
    </div>
    <div
      class="css-1j3ps0f"
    >
      <div
        class="css-1lqkgdc"
      >
        <p
          class="chakra-text css-1tafeqz"
          data-testid="header"
        >
          Login To Your Account
        </p>
        <form
          data-testid="form"
        >
          <div
            class="chakra-form-control css-1vtzp2s"
            role="group"
          >
            <label
              class="chakra-form__label css-1xn8xj5"
              data-testid="email-field"
              for="field-13"
              id="field-13-label"
            >
              Email address
            </label>
            <input
              class="chakra-input css-0"
              id="field-13"
              type="email"
            />
          </div>
          <div
            class="chakra-form-control css-1vtzp2s"
            role="group"
          >
            <label
              class="chakra-form__label css-1xn8xj5"
              data-testid="password-field"
              for="field-14"
              id="field-14-label"
            >
              Password
            </label>
            <input
              class="chakra-input css-0"
              id="field-14"
              type="password"
            />
          </div>
          <button
            class="chakra-button css-wkulga"
            data-testid="login-button"
            type="submit"
          >
            Log In
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
